Add FXIOS-8326, DISCO-2720 [v124] Record Category 2 impressions and clicks for Firefox Suggestions.

linabutler opened the pull request:
## :scroll: Tickets
[Jira ticket](https://mozilla-hub.atlassian.net/browse/FXIOS-8326)
[Github issue](https://github.com/mozilla-mobile/firefox-ios/issues/18449)

## :bulb: Description

This PR builds on #18273 to add Category 2 events for impressions and clicks on sponsored and non-sponsored suggestions. There are three differences between these new metrics, and the existing Category 2.5 `fx_suggest` metrics:

* These new metrics are events, submitted in the `events` ping. The existing metrics are submitted in a separate `fx-suggest` ping.
* The new events include the `client_id`, so we can use them for Nimbus experiment analysis, and correlate them with metrics for other features. The existing `fx-suggest` ping does _not_ include the `client_id`, since it contains Category 2.5 data.
* We record the new `search_result_impression` event for engaged and abandoned search sessions, and note the engagement state in the extra `search_result_impression.engagement` key. This gives us a more accurate picture of the impact of the new suggestion types. The existing `fx-suggest` ping is only sent for impressions during an engaged search session, because this ping is used for partner reporting and revenue analysis, and we don't receive any revenue for abandonments.

## :pencil: Checklist
You have to check all boxes before merging
- [ ] Filled in the above information (tickets numbers and description of your work)
- [ ] Updated the PR name to follow our [PR naming guidelines](https://github.com/mozilla-mobile/firefox-ios/wiki/Pull-Request-Naming-Guide)
- [ ] Wrote unit tests and/or ensured the tests suite is passing
- [ ] When working on UI, I checked and implemented accessibility (minimum Dynamic Text and VoiceOver)
- [ ] If needed I updated documentation / comments for complex code and public methods

linabutler commented:
# Request for data collection review form

1) What questions will you answer with this data?

* Are users shown sponsored and non-sponsored suggestions when they interact with the awesomebar?
* Do users click / tap on these suggestions after they're shown?
* Do users abandon their searches after seeing these suggestions, or complete their searches by tapping on these or other suggestions?
* What is the baseline for sponsored and non-sponsored impressions and clicks on mobile?

2) Why does Mozilla need to answer these questions?  Are there benefits for users? Do we need this information to address product or business requirements?

Product and business requirements, to validate the Enhanced Cross-Platform Suggest experiment.

3) What alternative methods did you consider to answer these questions? Why were they not sufficient?

Impressions for engaged awesomebar searches, and clicks, are currently collected in the `fx-suggest` ping.

However, the `fx-suggest` ping intentionally omits the `client_id`, because it contains Category 2.5 data. Our experiment analysis tooling relies on the `client_id`, so the metrics collected in the `fx-suggest` ping are not suitable for experimentation.

4) Can current instrumentation answer these questions?

No.

5) List all proposed measurements and indicate the category of data collection for each measurement, using the [Firefox data collection categories](https://wiki.mozilla.org/Data_Collection) found on the Mozilla wiki.

This PR adds the following measurements to the the `events` ping:

| Measurement                           | Description                                                                                                                                                                                                                                        | Data Collection Category | Tracking Bug # |
|---------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|--------------------------|----------------|
| `awesomebar.search_result_tap`        | Adding `sponsored-amp-suggestion` and `non-sponsored-wikipedia-suggestion` as possible values for the extra `type` key.                                                                                                                            | 2                        | 18449          |
| `awesomebar.search_result_impression` | A sponsored (`sponsored-amp-suggestion`) or non-sponsored (`non-sponsored-wikipedia-suggestion`) was visible when the user finished interacting with the awesomebar. The extra `engagement` key specifies how the user finished their interaction. | 2                        | 18449          |

6) Please provide a link to the documentation for this data collection which describes the ultimate data set in a public, complete, and accurate way.

This collection is Glean so is documented [in the Glean Dictionary](https://dictionary.telemetry.mozilla.org).

7) How long will this data be collected?  Choose one of the following:

This data will be collected permanently. The Suggest team will be responsible for the permanent collections.

8) What populations will you measure?

US-only users who are enrolled in the Enhanced Cross-Platform Suggest experiment.

9) If this data collection is default on, what is the opt-out mechanism for users?

These collections are Glean. The opt-out can be found in Settings.

10) Please provide a general description of how you will analyze this data.

General reporting on Looker and Redash, as well as experiment-specific analyses.

11) Where do you intend to share the results of your analysis?

The results will be shared internally with the Suggest team.

12) Is there a third-party tool (i.e. not Glean or Telemetry) that you are proposing to use for this data collection?

No.

linabutler commented:
/cc @adudenamedruby

adudenamedruby commented on firefox-ios/Client/Telemetry/TelemetryWrapper.swift:
```diff
@@ -1875,14 +1887,40 @@ extension TelemetryWrapper {
+            // Submit a separate `fx-suggest` ping for this impression.
+            // These pings do not include the `client_id`, and we only submit
+            // them for engaged search sessions.
+            guard didAbandonEngagement else {
+                break
+            }
```
```suggestion
            guard didAbandonEngagement else { break }
```

nit

adudenamedruby commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -4164,6 +4164,32 @@ awesomebar:
+          a search term.
+        type: boolean
+    bugs:
+      - TODO
```
```suggestion
      - https://mozilla-hub.atlassian.net/browse/FXIOS-8326
```

adudenamedruby commented on firefox-ios/Client/metrics.yaml:
```diff
@@ -4164,6 +4164,32 @@ awesomebar:
+    bugs:
+      - TODO
+    data_reviews:
+      - TODO
```
```suggestion
      - https://github.com/mozilla-mobile/firefox-ios/pull/18452
```

linabutler commented:
Hi @adudenamedruby! I changed a couple of things, and wanted to re-request data and code review from you:

* I removed the `engagement_abandoned` extra key from `awesomebar.search_result_impression`, because...
* I added `awesomebar.engagement` and `awesomebar.abandoned` events, to record when the user completes or abandons their search session.

(I updated the description and the data review request form to capture these changes. Apologies for the churn!)

This should give us the denominators we need for our experiment's [feature metrics](https://docs.google.com/document/d/19G70_G09wFc4cZ-8vYnC-dmLdfNQMoCQrbaaTtOdO8A/edit#heading=h.vvlndj64b2zj).

linabutler commented on firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:
```diff
@@ -2383,20 +2383,31 @@ extension BrowserViewController: SearchViewControllerDelegate {
-        let visibleSuggestionsTelemetryInfo = searchViewController.visibleSuggestionsTelemetryInfo
-        for info in visibleSuggestionsTelemetryInfo {
-            trackEngagedSearchTelemetry(visibleSuggestionInfo: info)
+        switch searchEngagementState {
```
I tried to DRY this out with something like:

```swift
var method: TelemetryWrapper.EventMethod
switch searchEngagementState {
case .engaged:
    method = .engagement

case .abandoned:
    method = .abandonment

default:
    return
}

let visibleSuggestionsTelemetryInfo = searchViewController.visibleSuggestionsTelemetryInfo
visibleSuggestionsTelemetryInfo.forEach { trackVisibleSuggestion(telemetryInfo: $0) }
TelemetryWrapper.gleanRecordEvent(category: .action, method: method, object: .locationBar)
```

...But I think that actually looks clunkier than duplicating the two extra lines in the cases. I'd be happy to change it if you prefer it, though!

adudenamedruby replied:
yeh.... I DRY at 3+. Two repetitions are whatever. lol

linabutler commented on firefox-ios/Client/Frontend/Browser/BrowserViewController/Views/BrowserViewController.swift:
```diff
@@ -2405,6 +2416,7 @@ extension BrowserViewController: SearchViewControllerDelegate {
                     TelemetryWrapper.EventValue.fxSuggestionTelemetryInfo.rawValue: telemetryInfo,
                     TelemetryWrapper.EventValue.fxSuggestionPosition.rawValue: position,
                     TelemetryWrapper.EventValue.fxSuggestionDidTap.rawValue: didTap,
+                    TelemetryWrapper.EventValue.fxSuggestionDidAbandonEngagement.rawValue: didAbandonEngagement,
```
`didAbandonSearchSession`? `isAbandonment`? Some other, better name to tie this to the engagement and abandonment events?

adudenamedruby replied:
`didAdandonSearchSession` sounds descriptively good to me!

linabutler replied:
Great! I pushed a second commit renaming `SearchEngagementState` to `SearchSessionState`, to be more consistent with this.

adudenamedruby reviewed (approved):
LGTM.

Also r+ on data review. Same applies as before, so it's still good.

linabutler commented:
@adudenamedruby Alrighty, this should be ready to go! Would you mind merging it for me, please? 😊 I tested locally, and verified that the new events are sent in all the scenarios we expect.
